components/engine_traits/src/table_properties.rs

```diff
@@ -25,4 +39,52 @@ pub trait TablePropertiesExt: CFHandleExt {
     }
 }
 
-pub trait TablePropertiesCollection {}
+pub trait TablePropertiesCollection<I, PKey, P, UCP>
+where
+    I: TablePropertiesCollectionIter<PKey, P, UCP>,
+    PKey: TablePropertiesKey,
+    P: TableProperties<UCP>,
+    UCP: UserCollectedProperties,
+{
+    fn iter(&self) -> I;
+
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+pub trait TablePropertiesCollectionIter<PKey, P, UCP>
+where
+    Self: Iterator<Item = (PKey, P)>,
+    PKey: TablePropertiesKey,
+    P: TableProperties<UCP>,
+    UCP: UserCollectedProperties,
+{
+}
+
+pub trait TablePropertiesKey
+where
+    Self: Deref<Target = str>,
+{
+}
```

Comment:
why not write it like

```rust
pub trait TablePropertiesKey: Deref<Target = str> {}
```

Reply:
Mostly for stylistic consistency with nearby types that contain lots of where clauses.